File: 1463. Cherry Pickup II/cherryPickup.cpp
class Solution
{
public:
    int m, n;
    int t[71][71][71];
    int solve(vector<vector<int>> &grid, int row, int c1, int c2)
    {
        if (row >= n)
            return 0; // robot reached the last row;
        if (t[row][c1][c2] != -1)
            return t[row][c1][c2];
        int cherry = 0;
        cherry += grid[row][c1]; // if both are at the same cell take one
        if (c1 != c2)
        {
            cherry += grid[row][c2]; // cells are diff then take both;
        }

        int ans = 0;
        for (int i = -1; i <= 1; i++)
        {
            for (int j = -1; j <= 1; j++)
            {
                int nRow = row + 1;
                int nC1 = c1 + i;
                int nC2 = c2 + j;
                if (nC1 >= 0 && nC1 < m && nC2 >= 0 && nC2 < m)
                {
                    ans = max(ans, solve(grid, nRow, nC1, nC2));
                }
            }
        }

        return t[row][c1][c2] = cherry + ans;
    }
    int cherryPickup(vector<vector<int>> &grid)
    {
        n = grid.size();
        m = grid[0].size();
        memset(t, -1, sizeof(t));
        return solve(grid, 0, 0, m - 1);
    }
};

File: 1463. Cherry Pickup II/cherryPickup.java
class Solution {
    public int[][][] t;
    public int m, n;

    public int solve(int[][] grid, int row, int c1, int c2) {
        if (row >= n) {
            return 0; // Robot reached the last row
        }

        if (t[row][c1][c2] != -1) {
            return t[row][c1][c2];
        }

        int cherry = grid[row][c1]; // If both robots are at the same cell, take one cherry
        if (c1 != c2) {
            cherry += grid[row][c2]; // If robots are at different cells, take both cherries
        }

        int ans = 0;
        for (int i = -1; i <= 1; i++) {
            for (int j = -1; j <= 1; j++) {
                int nRow = row + 1;
                int nC1 = c1 + i;
                int nC2 = c2 + j;
                if (nC1 >= 0 && nC1 < m && nC2 >= 0 && nC2 < m) {
                    ans = Math.max(ans, solve(grid, nRow, nC1, nC2));
                }
            }
        }

        return t[row][c1][c2] = cherry + ans;
    }

    public int cherryPickup(int[][] grid) {
        n = grid.length;
        m = grid[0].length;
        t = new int[71][71][71];
        for (int i = 0; i < 71; i++) {
            for (int j = 0; j < 71; j++) {
                Arrays.fill(t[i][j], -1);
            }
        }
        return solve(grid, 0, 0, m - 1);
    }
}